Add WebSocket permessage-deflate

AntonRutkevich opened the pull request:
Resolves #1733 

Hi! 

Here's the draft of the ["permessage-deflate"](https://tools.ietf.org/html/rfc7692#section-7) compression for WebSocket. 

It has at least one unresolved issue discussed in https://github.com/square/okio/issues/570. 
It would be great to get your feedback on the overall approach. 

Thanks!

ming13 commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -149,6 +149,8 @@ open class OkHttpClient internal constructor(
   private val readTimeout: Int = builder.readTimeout
   private val writeTimeout: Int = builder.writeTimeout
   private val pingInterval: Int = builder.pingInterval
+  private val webSocketCompression: Boolean = builder.webSocketCompression
```
Is it possible to have a number of compression algorithms? In such case it would be advisable to have `enum` here, like:

```kotlin
enum class Compression { PermessageDeflate, None }
```

AntonRutkevich replied:
AFAIK, "permessage-deflate" is the only widely used. Don't see a reason to overcomplicate ATM.

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt:
```diff
@@ -0,0 +1,49 @@
+ * Applies post-deflate and pre-inflate buffer transformations
+ * described in [rfc7692#section-7.2](https://tools.ietf.org/html/rfc7692#section-7.2)
+ */
+object CompressionCorrector {
```
I would suggest to make it a `class` to avoid having an unnecessary singleton implementation (which will be in memory all the time). Plus this class is stateless and doesn’t seem to need instance in the first place.

AntonRutkevich replied:
Agree on the unnecessary `INSTANCE`. 
As this class is essentially 2 stateless functions, I've removed the `object` in favor of top level functions.

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt:
```diff
@@ -0,0 +1,49 @@
+    writeInt(FOUR_OCTETS_TO_ADD_BEFORE_INFLATION)
+  }
+
+  private fun endsWithEmptyDeflateBlock(buffer: Buffer): Boolean = buffer.rangeEquals(
```
Do you need it anywhere else? Seems like a candidate for inlining into `applyPostDeflate`.

AntonRutkevich replied:
I would rather leave it as is for the sake of readability. Let's wait for other opinions if any.

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt:
```diff
@@ -0,0 +1,49 @@
+object CompressionCorrector {
+
+  private val EMPTY_DEFLATE_BLOCK = "000000ffff".decodeHex()
+  private const val LAST_OCTETS_TO_REMOVE_AFTER_DEFLATION = 4
```
`LAST_OCTETS_COUNT_TO_REMOVE_AFTER_DEFLATION`?

Damn, I’m practically hearing how this name screams at me with these caps letters...

AntonRutkevich replied:
Done

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt:
```diff
@@ -0,0 +1,49 @@
+  private val EMPTY_DEFLATE_BLOCK = "000000ffff".decodeHex()
+  private const val LAST_OCTETS_TO_REMOVE_AFTER_DEFLATION = 4
+
+  private const val FOUR_OCTETS_TO_ADD_BEFORE_INFLATION = 0x0000ffff
```
`OCTETS_TO_ADD_BEFORE_INFLATION`?

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,32 @@
+
+class MessageInflater : Closeable {
+  private val source = Buffer()
+  private val inflater = Inflater(true /* omit zlib header */)
```
Thank you for this comment.

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,32 @@
+      try {
+        writeAll(inflaterSource)
+      } catch (e: EOFException) {
+        e.printStackTrace()
```
Pants on fire!

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,32 @@
+  @Throws(IOException::class)
+  override fun close() {
+    inflaterSource.close()
+  }
```
```kotlin
override fun close() = inflaterSource.close()
```
?

AntonRutkevich replied:
Done

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.java:
```diff
@@ -212,6 +226,13 @@ void writeMessageFrame(int formatOpcode, long byteCount, boolean isFirstFrame,
+  @Override
+  public void close() throws IOException {
+    if (messageDeflater != null) {
+      messageDeflater.close();
```
```kotlin
messageDeflater?.close()
```
[Ah, that’s Java](https://66.media.tumblr.com/5ea028443f6ad88bf6af6eb24f1294be/tumblr_mh2htyUWqO1rzupqxo1_500.png).

ming13 commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,57 @@
+import okhttp3.Response
+import java.io.IOException
+
+data class WebSocketOptions(
```
Please make sure you’ll cover all this with tests.

AntonRutkevich replied:
Well, sure, just gathering feedback on the approach now.

JakeWharton commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.java:
```diff
@@ -128,12 +132,25 @@ private void readHeader() throws IOException {
     }
 
-    boolean reservedFlag1 = (b0 & B0_FLAG_RSV1) != 0;
+    boolean compressedFlag = (b0 & B0_FLAG_RSV1) != 0;
```
I don't think we should rename this. The meaning of the flag only indicates compressed when the extension is specified. And the check below should still throw when the extension is not specified and this bit is set.

AntonRutkevich replied:
Makes sense. Fixed.

AntonRutkevich commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.java:
```diff
@@ -221,7 +241,13 @@ void writeMessageFrame(int formatOpcode, long byteCount, boolean isFirstFrame,
+            messageDeflater.deflate(source));
+      }
+
+      buffer.write(uncompressedSource, uncompressedSource.size());
```
Just realized that compressing the whole message here, inside FrameSink.write() is not correct, as the method guarantees to write exactly `byteCount` number of bytes. 

Besides, there is logic in `RealWebSocket.send(ByteString data, int formatOpcode)` which checks for queue overflow. With this change, the logic there will count the number of _uncompressed_ bytes, which seems to be incorrect as it differs from the number of bytes to be sent on the wire. 
So, ideally the message data should be compressed by the time it is put in `messageAndCloseQueue` in the above mentioned `send` method. However, the method does not run on the writer thread, and we don't know the thread it is running on.  

Thinking how to resolve this... Suggestions are welcome!

AntonRutkevich replied:
If we put aside the queue overflow checks, we could put the deflation code in the `RealWebSocket.writeOneFrame` method, like that: 

```
ByteString data = ((Message) messageOrClose).data;

// deflate data here and pass it down to newMessageSink

BufferedSink sink = Okio.buffer(writer.newMessageSink(
    ((Message) messageOrClose).formatOpcode, data.size()));
sink.write(data);
sink.close();
```

However, it does not seem to be a good solution.

AntonRutkevich replied:
On second thought, `MAX_QUEUE_SIZE` is 16MB. I suppose it could be 12MB or 20MB or any other reasonable number. So if we treat `queueSize` as the number of _uncompressed_ bytes in the queue, the above solution with `RealWebSocket.writeOneFrame` could work.

AntonRutkevich replied:
Implemented this approach.

swankjesse replied:
Good call.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageDeflater.kt:
```diff
@@ -0,0 +1,31 @@
+package okhttp3.internal.ws
```
nit: copyright header

AntonRutkevich replied:
Done

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageDeflater.kt:
```diff
@@ -0,0 +1,31 @@
+
+class MessageDeflater(private val contextTakeover: Boolean) : Closeable {
+  private val sink = Buffer()
+  private val deflater = Deflater(Deflater.BEST_COMPRESSION, true /* omit zlib header */)
```
are you sure you want `BEST_COMPRESSION`? Typically  DEFAULT_COMPRESSION is appropriate for network where we need to balance time compressing vs. time transmitting

AntonRutkevich replied:
Did not run any benchmarks so far. 
OK, will set it to default compression to be on the safe side. Will raise additional PR if benchmarks show  `BEST_COMPRESSION` fits better.

swankjesse replied:
Benchmarking won’t show you much because we’re just deciding how to trade off CPU+memory for network.

swankjesse replied:
(benchmarking will tell you that your test environment has lots of network and so any compression at all is wasted)

AntonRutkevich replied:
OK, I'm totally fine with `DEFAULT_COMPRESSION`.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,36 @@
+package okhttp3.internal.ws
```
nit: copyright

AntonRutkevich replied:
Done

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,36 @@
+
+class MessageInflater : Closeable {
+  private val source = Buffer()
+  private val inflater = Inflater(true /* omit zlib header */)
```
nice comment!

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,36 @@
+      // Instead, we ensure that all bytes from source have been processed by inflater.
+      while (true) {
+        inflaterSource.read(this, Long.MAX_VALUE)
+        if (inflater.totalIn.toLong() == totalBytesToInflate) {
```
perfecto

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,36 @@
+
+  @Throws(IOException::class)
+  fun inflate(deflated: Buffer): Buffer {
+    source.writeAll(deflated)
```
maybe before this add `require(source.size == 0)` ?

AntonRutkevich replied:
Done

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,36 @@
+    source.writeAll(deflated)
+    totalBytesToInflate += source.size
+
+    return Buffer().apply {
```
should this method write the inflated bytes back to the caller’s buffer?

swankjesse replied:
It’s already destructive and that saves an allocation

AntonRutkevich replied:
Done

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,36 @@
+  private val inflater = Inflater(true /* omit zlib header */)
+  private val inflaterSource = InflaterSource(source, inflater)
+
+  private var totalBytesToInflate = 0L
```
should this be an Int ?

swankjesse replied:
(if we inflate more than 2 GiB of data, I assume totalIn wraps around?)

AntonRutkevich replied:
Checked the javadocs for `totalIn` again:

```
     * Since the number of bytes may be greater than
     * Integer.MAX_VALUE, the {@link #getBytesRead()} method is now
     * the preferred means of obtaining this information.</p>
```

So, migrated to `getBytesRead` instead of moving this to int.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.java:
```diff
@@ -163,6 +177,11 @@ public RealWebSocket(Request request, WebSocketListener listener, Random random,
     };
   }
 
+  public RealWebSocket(Request request, WebSocketListener listener, Random random,
```
I’m looking forward to using default parameters

AntonRutkevich replied:
Yep will simplify a bit.

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -303,6 +312,8 @@ open class OkHttpClient internal constructor(
     internal var readTimeout: Int = 10000
     internal var writeTimeout: Int = 10000
     internal var pingInterval: Int = 0
+    internal var webSocketCompression: Boolean = false
```
this should be true by default?

AntonRutkevich replied:
Setting it to true will silently enable compression for existing web socket users (if their servers supports it).
If you're sure it's appropriate, I can update it to be true by default.

swankjesse replied:
Lets do 2 PRs – first to check this all in, off by default. And a second PR that only flips the default.

OkHttp’s design objective includes this phrase “OkHttp is an HTTP client that’s efficient by default”. As far as I’m concerned that means compression is enabled by default.

swankjesse replied:
ALSO – can you think of any reason why end-users would want to disable compression? I think we should always advertise support for it and let the server decide.

swankjesse replied:
(ie. remove the fiddly configuration flags until we’re presented with a concrete use case justifying them.)

AntonRutkevich replied:
I personally don't see any reasons why we should keep compression off. My only concern was the silent default change. 
Sure, let's go with your approach.

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -303,6 +312,8 @@ open class OkHttpClient internal constructor(
     internal var writeTimeout: Int = 10000
     internal var pingInterval: Int = 0
+    internal var webSocketCompression: Boolean = false
+    internal var webSocketContextTakeover: Boolean = false
```
this should be true by default?

AntonRutkevich replied:
Same as about webSocketCompression above.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.java:
```diff
@@ -261,4 +293,11 @@ private void readMessage() throws IOException {
+  @Override
+  public void close() throws IOException {
+    if (messageInflater != null) {
+      messageInflater.close();
```
nice

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.java:
```diff
@@ -493,13 +541,20 @@ boolean writeOneFrame() throws IOException {
+
+        ByteString data = uncompressedData;
+        if (messageDeflater != null) {
+          data = CompressionCorrector.applyPostDeflate(
```
Combine CompressionCorrector and MessageDeflater. They’re useless without each other, right?

AntonRutkevich replied:
Done, same for MessageInflater. CompressionCorrector is now gone.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,56 @@
+    @JvmStatic
+    fun parseServerResponse(response: Response): WebSocketOptions =
+        response.header("Sec-WebSocket-Extensions")
+            ?.takeIf { it.contains("permessage-deflate") }
```
I’m tempted to ask you to use less Kotlin and more if/else

AntonRutkevich replied:
Sure. Any details on why, so that I can better follow in the future?

swankjesse replied:
I find this method in particular a little Perl-ish. I need to study this to learn what it’s really doing. Maybe some day I’ll be good enough at Kotlin where I can read this naturally, but for now I think this code loses readability to be compact.

AntonRutkevich replied:
OK, no problem.
A bigger rewrite of the WebSocketOptions parsing is on its way, I've streamlined the code there.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,56 @@
+              val clientMaxWindowBits = parseClientMaxWindowBits(options)
+
+              if (!clientNoContextTakeover) {
+                if (clientMaxWindowBits != null
```
combine if blocks?

AntonRutkevich replied:
Done. 
Actually, I think I should further improve it to be [closer to RFC,](https://tools.ietf.org/html/rfc7692#section-5.2) eg supporting quoted parameter values and fallback options.
Will do.

AntonRutkevich replied:
Done

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.java:
```diff
@@ -131,9 +139,27 @@ private void readHeader() throws IOException {
+
+    if (reservedFlag1 && !compressionEnabled) {
+      throw new ProtocolException(
+          "Reserved flag rsv1 is not supported if compression is disabled.");
```
nice

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.java:
```diff
@@ -131,9 +139,27 @@ private void readHeader() throws IOException {
+    if (reservedFlag2 || reservedFlag3) {
       // Reserved flags are for extensions which we currently do not support.
-      throw new ProtocolException("Reserved flags are unsupported.");
+      throw new ProtocolException("Reserved flags rsv2 and rsv3 are unsupported.");
```
Nice.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.java:
```diff
@@ -131,9 +139,27 @@ private void readHeader() throws IOException {
+      throw new ProtocolException("Reserved flag rsv1 must not be set on control frames.");
+    }
+
+    if (reservedFlag1 && opcode == OPCODE_CONTINUATION) {
```
can you refactor this so we read reservedFlag1 only once?

AntonRutkevich replied:
Indeed :) Done.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.java:
```diff
@@ -131,9 +139,27 @@ private void readHeader() throws IOException {
+    }
+
+    if (compressionEnabled && opcode == OPCODE_TEXT || opcode == OPCODE_BINARY) {
+      readingCompressedMessage = reservedFlag1;
```
this is the 4th time we’ve read that flag!

AntonRutkevich replied:
Done

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.java:
```diff
@@ -53,12 +54,15 @@
   final Buffer buffer = new Buffer();
   final FrameSink frameSink = new FrameSink();
 
+  final boolean compressionEnabled;
```
shouldn’t this be a parameter of `writeMessageFrame` ? I don’t think WebSocketWriter wants to be limited to always compressed or always uncompressed. If that were the case the protocol wouldn’t have facilities for changing that on a per-message basis

AntonRutkevich replied:
You're right. 

Actually, rfc [suggests](https://tools.ietf.org/html/rfc7692#section-7.3) the trick where messages can stay uncompressed if their deflated versions are bigger in size **and no** "context takeover" is used. Implemented.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.java:
```diff
@@ -173,6 +179,9 @@ void writeMessageFrame(int formatOpcode, long byteCount, boolean isFirstFrame,
     if (isFinal) {
       b0 |= B0_FLAG_FIN;
     }
+    if (compressionEnabled && isFirstFrame) {
```
I think we just want a parameter like `boolean compressionEnabled`.

AntonRutkevich replied:
Yep. I went for `isCompressed`, looks like it better describes the fact that it's about this exact message, as opposed to some global state.

AntonRutkevich commented on okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt:
```diff
@@ -0,0 +1,42 @@
+ * @param deflated Buffer containing deflated data
+ */
+fun applyPostDeflate(deflated: Buffer) = deflated.apply {
+  if (endsWithEmptyDeflateBlock(this)) {
```
The rfc [suggests](https://tools.ietf.org/html/rfc7692#section-7.2.1) the following approach: 

```
   1.  Compress all the octets of the payload of the message using
       DEFLATE.

   2.  If the resulting data does not end with an empty DEFLATE block
       with no compression (the "BTYPE" bits are set to 00), append an
       empty DEFLATE block with no compression to the tail end.

   3.  Remove 4 octets (that are 0x00 0x00 0xff 0xff) from the tail end.
       After this step, the last octet of the compressed data contains
       (possibly part of) the DEFLATE header bits with the "BTYPE" bits
       set to 00.
```

Empty deflate block AFAIK is 5 bytes: `00 00 00 ff ff`. 
So, if the deflated data does not end with it, we first append these 5 bytes, and then remove last 4 of them, leaving the 1st of `00`. 
So, it's guaranteed that deflated messages sent on the wire end with `00`. 
Of inflation, we first append the previously removed 4 bytes of `00 00 ff ff`, so the resulting deflated data is guaranteed to end with empty deflate block `00 00 00 ff ff`. Only after the data is passed to the inflator. 

So, answering your question, if deflated application data ends with `00 00 00 ff ff`, it will be exactly the same before inflation.

In the `else` block here I just add `00` instead of adding `00 00 00 ff ff` and removing `00 00 ff ff`.

Does it make sense?

swankjesse replied:
I think I get it. The spec authors recognized that DEFLATE wasted four bytes so they created their own DEFLATE variant to avoid that.

SGtM

AntonRutkevich replied:
Exactly.

AntonRutkevich commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -255,6 +257,12 @@ open class OkHttpClient internal constructor(
+  /** Whether WebSocket compression was requested. */
+  open fun webSocketCompression(): Boolean = webSocketCompression
+
+  /** Whether WebSocket compression "context takeover" was requested. */
```
Sure!

yschimke replied:
"Everywhere you go you always take the whether with you, Everywhere you go you always take the whether, Take the whether, the whether with you." - Crowded House

AntonRutkevich commented on okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt:
```diff
@@ -0,0 +1,42 @@
+@file:JvmName("CompressionCorrector")
```
Done

AntonRutkevich commented on okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt:
```diff
@@ -0,0 +1,42 @@
+  writeInt(OCTETS_TO_ADD_BEFORE_INFLATION)
+}
+
+private fun endsWithEmptyDeflateBlock(buffer: Buffer): Boolean = buffer.rangeEquals(
```
Done

AntonRutkevich commented:
Thanks a lot for the review, @ming13 , @JakeWharton , @swankjesse !

Working on my broken git client)
Please disregard this PR till explicit notice.
I'll also change the status to non-draft once ready.

swankjesse commented:
This is one of the nicest PRs I’ve seen in a while. Thanks for your contribution @AntonRutkevich !

AntonRutkevich commented:
Thanks for your great review @swankjesse !

AntonRutkevich commented:
OK, now it's ready for review. 

1. Added WebSocketOptions parsing which is closer to RFC
2. Added tests

These changes are in the last 2 commits to make them easier to review. Once review is complete, I'll rebase it onto master branch, and squash into one commit.

@swankjesse , are there any other blockers for this to be merged (after review is done)?

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/MessageInflaterTest.kt:
```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
```
2019?

AntonRutkevich replied:
Fixed all

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketOptionsTest.kt:
```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
```
2019?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageDeflater.kt:
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2012 Square, Inc.
```
2019?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/MessageInflater.kt:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2012 Square, Inc.
```
2019?

AntonRutkevich commented:
@swankjesse 

> Can you mark the two new methods on OkHttpClient as internal? I’d like to restrict OkHttp 4.0 to ZERO new behavior + API changes, adding those in 4.1 and beyond.

Sure, done. 


I'm totally fine with it _on_ by default, but the trick with the header does not work here as nicely as with gzip, as far as I can tell.

For gzip, there is a way to disable it with a request header, with `Accept-Encoding: identity`. I could not find a way disable the websocket compression **with** "Sec-WebSocket-Extensions" header present. Maybe just something hacky like sending a non-existing compression algorithm, like `Sec-WebSocket-Extensions: none`, forcing the server utilize this part of [RFC](https://tools.ietf.org/html/rfc7692#section-5): 

```
   A server declining all offered PMCEs MUST NOT include any element
   with PMCE names.  If a server responds with no PMCE element in the
   "Sec-WebSocket-Extensions" header, both endpoints proceed without
   per-message compression once _the WebSocket Connection is
   established_.
```

But if we do so, this PR and the source code will be the only place to learn how to do it for okhttp users. It's non-standard too. 

Do you see a better option here?

swankjesse commented:
Turn compression on and see what breaks?

AntonRutkevich commented:
OK, no problem. 
So I'll basically remove all the new stuff from OkHttpClient.
And fix the options parsing to treat empty PCME list as "no compression".

AntonRutkevich commented:
Done in [da99030](https://github.com/square/okhttp/commit/da99030228412861ce2a92faeda8c59d840982a6).

Now compression is on by default.

avently commented:
@AntonRutkevich 
Hello and thank your for the feature you made!
I'm using okhttp from your fork since yesterday. Before that I was using 3.x version.
Interesting error appears in logs in onFailure() method:
```
source exhausted prematurely
```
Didn't see this error before (my app have been working more than a year) so it's probably related to your PR. Could you check on your side?
For the info: this error shows itself in two different websockets from two different crypto exchanges so the websocket data can't be wrong. I got tens of the errors per day.

AntonRutkevich commented:
Hello @avently ,

Your error might definitely be related to this PR. I've seen it during development, when I was trying to get all these buffers work correctly.

Could you please share more details:
1. How often does the issue arise? 10s of errors per day, how much is it compared to the daily user count (just the order of magnitude)? Or, what percentage of users gets affected by this daily?
2. Do you have any idea about the pattern or any other details besides the error message? Message size? Common OS version, common device model, else? Could you reproduce it locally?

From my side, I've integrated a version of this PR rebased onto okhttp 3.13.1 (as I didn't want to include pre-release 4 version into a production app), and we'll be using it internally during at least next 1-2 weeks. We do collect error logs and other metrics, so I'll let you know if I find something.

avently commented:
> How often does the issue arise? 

6-7 erros per hour. Sometimes 2-3 per minute and then silence for an hour. I don't use the fork in production so it happens only on two of my devices: PC with JDK 8 and Android 7.1.

> Do you have any idea about the pattern or any other details besides the error message?

I have no idea how your code works (didn't read it) but maybe there is somewhere Integer.MAX_VALUE instead of Long.MAX_VALUE in bytes size. Some websocket messages can be larger than one frame's size.

> Payload size of the message that failed?

How to debug it? I only can duplicate websocket without encryption and see what the frame was when it's failed on encrypted websocket. 

Looks like a have to debug your changes if you can't reproduce the error.

AntonRutkevich commented:
Thanks for the details!

> 6-7 erros per hour.
That's a lot. Then the chances are high I'll see it once we run it in the staging env internally. 

> I have no idea how your code works (didn't read it) but maybe there is somewhere Integer.MAX_VALUE instead of Long.MAX_VALUE in bytes size. Some websocket messages can be larger than one frame's size.
Sure, I was just wandering if you happened to have any ideas based on knowledge of your web socket traffic patterns, or something else local to your case.

> How to debug it? I only can duplicate websocket without encryption and see what the frame was when it's failed on encrypted websocket.
I was thinking you could know it in advance, for example. It can also be debugged in wireshark. 

Anyway, I'll try to take a closer look next week.

avently commented:
New info.

I added some logs and try-catch to the function inflate(). Parsed websocket message is correct even after exception is catched:
```
fun inflate(buffer: Buffer): Buffer {
    require(source.size == 0L)

    source.writeAll(buffer)
    println("LALAL WEBSOCKET inflate WRITEALL, source size " + source.size)
    // Adds 4 bytes
    source.writeInt(OCTETS_TO_ADD_BEFORE_INFLATION)

    totalBytesToInflate += source.size
    println("LALAL WEBSOCKET inflate totalbytestoinflate " + totalBytesToInflate)

    return buffer.apply {
      // We cannot read all, as the source does not close.
      // Instead, we ensure that all bytes from source have been processed by inflater.
      try {
        while (true) {
          println("LALAL WEBSOCKET inflate read bytesBefore " + inflater.bytesRead +
                  " value " + inflaterSource.read(this, Long.MAX_VALUE) +
                  " bytesRead " + inflater.bytesRead)
          if (inflater.bytesRead == totalBytesToInflate) {
            break
          }
        }
      } catch (e: Exception) {
        e.printStackTrace()
        println("LALAL WEBSOCKET inflate failed, bytes read " + inflater.bytesRead)
      }
    }
  }
```


And this is what I see in the logs:

Normal situation, 31 bytes + 4 bytes (from OCTETS_TO_ADD_BEFORE_INFLATION) were read.
```
LALAL WEBSOCKET inflate WRITEALL, source size 31
LALAL WEBSOCKET inflate totalbytestoinflate 18016
LALAL WEBSOCKET inflate read bytesBefore 17981 value 156 bytesRead 18016
```

Bad situation
```
LALAL WEBSOCKET inflate WRITEALL, source size 1507
LALAL WEBSOCKET inflate totalbytestoinflate 7146
LALAL WEBSOCKET inflate read bytesBefore 5635 value 5248 bytesRead 7142
java.io.EOFException: source exhausted prematurely
        at okio.InflaterSource.read(InflaterSource.kt:75)
        at okhttp3.internal.ws.MessageInflater.inflate(MessageInflater.kt:56)
        at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:254)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:117)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:321)
        at okhttp3.internal.ws.RealWebSocket$1.onResponse(RealWebSocket.java:253)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.kt:138)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
LALAL WEBSOCKET inflate failed, bytes read 7146
```
In the end you see that source was read fully. The difference here is the last 4 bytes. For some reason these 4 bytes wasn't read from first iteration. Then Okio decided that source is exhausted prematurely but read another 4 bytes.

And this is another example of normal situation but with more bytes:

```
LALAL WEBSOCKET inflate WRITEALL, source size 29222
LALAL WEBSOCKET inflate totalbytestoinflate 259013
LALAL WEBSOCKET inflate read bytesBefore 229787 value 8192 bytesRead 232122
LALAL WEBSOCKET inflate read bytesBefore 232122 value 8192 bytesRead 234303
LALAL WEBSOCKET inflate read bytesBefore 234303 value 8192 bytesRead 236446
LALAL WEBSOCKET inflate read bytesBefore 236446 value 5822 bytesRead 237975
LALAL WEBSOCKET inflate read bytesBefore 237975 value 44 bytesRead 237979
LALAL WEBSOCKET inflate read bytesBefore 237979 value 2326 bytesRead 238612
LALAL WEBSOCKET inflate read bytesBefore 238612 value 8192 bytesRead 240758
LALAL WEBSOCKET inflate read bytesBefore 240758 value 8192 bytesRead 242939
LALAL WEBSOCKET inflate read bytesBefore 242939 value 8192 bytesRead 245109
LALAL WEBSOCKET inflate read bytesBefore 245109 value 4071 bytesRead 246171
LALAL WEBSOCKET inflate read bytesBefore 246171 value 4121 bytesRead 247373
LALAL WEBSOCKET inflate read bytesBefore 247373 value 8192 bytesRead 249631
LALAL WEBSOCKET inflate read bytesBefore 249631 value 8192 bytesRead 251931
LALAL WEBSOCKET inflate read bytesBefore 251931 value 8192 bytesRead 254222
LALAL WEBSOCKET inflate read bytesBefore 254222 value 577 bytesRead 254363
LALAL WEBSOCKET inflate read bytesBefore 254363 value 7615 bytesRead 256454
LALAL WEBSOCKET inflate read bytesBefore 256454 value 8192 bytesRead 258691
LALAL WEBSOCKET inflate read bytesBefore 258691 value 1219 bytesRead 259013
```

Okio divides the source into chunks with different sizes (size shown after `value`), max is 8192 as defined inside Segment.SIZE variable. 

Bad situation with more bytes:
```
LALAL WEBSOCKET inflate WRITEALL, source size 3810
LALAL WEBSOCKET inflate totalbytestoinflate 2489316
LALAL WEBSOCKET inflate read bytesBefore 2485502 value 8192 bytesRead 2487047
LALAL WEBSOCKET inflate read bytesBefore 2487047 value 8192 bytesRead 2488688
LALAL WEBSOCKET inflate read bytesBefore 2488688 value 3050 bytesRead 2489312
java.io.EOFException: source exhausted prematurely
        at okio.InflaterSource.read(InflaterSource.kt:75)
        at okhttp3.internal.ws.MessageInflater.inflate(MessageInflater.kt:56)
        at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:254)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:117)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:321)
        at okhttp3.internal.ws.RealWebSocket$1.onResponse(RealWebSocket.java:253)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.kt:138)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
LALAL WEBSOCKET inflate failed, bytes read 2489316
```

From my point of view your code is correct. Why Okio throws exception but successfully reads the source? Who knows.

AntonRutkevich commented:
Hmm.. 
I have a guess.

In 
```
// Adds 4 bytes
source.writeInt(OCTETS_TO_ADD_BEFORE_INFLATION)
```
we rely on the fact that the writer (server) has stripped last 4 octets 0x0000ffff from the compressed data (as we do in MessageDeflater). 
If this is not the case, it might be possible that we get a data ending with `EMPTY_DEFLATE_BLOCK = 0x000000ffff` (see MessageDeflater), add our 4 octets of `0x0000ffff`.
Then okio stops reading at the empty deflate block, treating it as the end of the input, but we still have our `0x0000ffff` left in the Buffer. So, okio throws the EOFException. 

If this is the case, then fix like 
```
if (!endsWithEmptyDeflateBlock()) {
 source.writeInt(OCTETS_TO_ADD_BEFORE_INFLATION)
}
... 
// As in MessageInflater
private fun Buffer.endsWithEmptyDeflateBlock(): Boolean = rangeEquals(
      size - EMPTY_DEFLATE_BLOCK.size, EMPTY_DEFLATE_BLOCK)
```
should help. 

Again, I'm not sure it's the case, as I'm not a Deflate RFC pro. 

@avently Could you try this one?

If this does not work, then the `println()` of last 6 bytes of the block before exception would be useful. It can be obtained with `Buffer.getByte(int pos)`.

avently commented:
endsWithEmptyDeflateBlock() doesn't work.
The info you asked:
```
for (i in buffer.size - 9 until buffer.size) {
          println("$i is " + buffer[i])
        }
```
9455 is 49
9456 is 54
9457 is 48
9458 is 57
9459 is 57
9460 is 125
9461 is 93
9462 is 125
9463 is 93



5733 is 34
5734 is 58
5735 is 55
5736 is 54
5737 is 48
5738 is 125
5739 is 93
5740 is 125
5741 is 93


6873 is 56
6874 is 52
6875 is 56
6876 is 57
6877 is 53
6878 is 125
6879 is 93
6880 is 125
6881 is 93

The last four bytes is always 125, 93, 125, 93. This "snapshot" was taken in  catch() block

AntonRutkevich commented:
Thanks for the logs! 

I would expect every deflated message to end with `0x00`, so that we add `0x0000ffff` to get the empty deflate block. Looks like it's not always the case for both successful and failed ws messages. 

Do you know what's the web socket implementation on the server side?

BTW, I still cannot reproduce it locally.

avently commented:
> Do you know what's the web socket implementation on the server side?

Nope.

Attaching a sample app. It was initially made for reproducing another bug so it has more code than needed it the situation.

Change a version of okhttp lib in the build.gradle to your local version. I use the version 4.0.0-SNAPSHOT (just named it like that for myself).

Run with ./gradlew build runOkhttp

Wait a liitle bit and you'll see a crash.

[sample.zip](https://github.com/square/okhttp/files/3252749/sample.zip)

AntonRutkevich commented:
Thanks a lot for the sample! 

I was able to reproduce the issue, actually your sample throws the exception almost immediately.

While I still don't have a proper solution, here is what I found:
1) Looks like the backend implementation you are listening to is not following the ws permessage-deflate spec directly ([7.2.1](https://tools.ietf.org/html/rfc7692#section-7.2.1)) as there are messages which do not end with 0x00. Still, I do confirm that there seems to be no explicit correlation between that and the exception: 0x00 messages produce the exception sometimes, and non-0x00 messages are parsed correctly sometimes.
2) Again, I do confirm that if `inflater.bytesRead == totalBytesToInflate - 4` after the `read` call, the exception is guaranteed to be thrown. 
This happens because on the next iteration inflater source

- tries to inflate these `0x0000ffff` bytes, but gets 0 bytes inflated
- so, it runs another cycle of the inner `while` loop, where the buffer is already exhausted, reads 0 bytes again -> throws the exception.

3) I have once been able to reproduce this with 5, not 4 bytes left to read: `0x000000ffff`. The message ended with `0x38e061042c090000`.
4) Also, the traffic pattern of the backend you are using is worth mentioning. 
Here is the example sizes of the first messages (159, 64, 29, 10, 16, 1755, 105657, 279888, 90, 32, 46, ...). So first some small meta info, then 1-2 huge snapshots, and then more smaller updates. 
From my experience, these huge messages throw the exception most often.

Here is what I tried with no success (in pseudocode): 

```java
// Only write 4 bytes if it will make the message end with empty deflate block. 
if (source.endsWith(0x00)) {
  source.writeInt(0x0000ffff)
}
```

```java
if (source.endsWith(0x00)) {
  source.writeInt(0x0000ffff)
} else {
  // Write full 5 bytes of empty deflate block with no compression: 0x000000ffff.
  source.writeByte(0x00)
  source.writeInt(0x0000ffff)
}
```

The only hack that seems to solve the issue I see now is 
```java
if (inflater.bytesRead == totalBytesToInflate || (inflater.bytesRead == totalBytesToInflate - 4)) {
  break
}
```
but it's not much better than try-catch of EOF with silent ignore. 

For now, I wasn't able to find the reason, why these 4 bytes are left unread, which seems to be the root cause.

avently commented:
I have 
```
if (!source.endsWithEmptyDeflateBlock()) {
      source.writeInt(OCTETS_TO_ADD_BEFORE_INFLATION)
}
```

with 
```
if (inflater.bytesRead == totalBytesToInflate || (inflater.bytesRead == totalBytesToInflate - 4)) {
     break
}
```

And it doesn't help me to prevent crashes. So only try-catch block works for now

AntonRutkevich commented:
Remove the 
```
if (!source.endsWithEmptyDeflateBlock()) {
      source.writeInt(OCTETS_TO_ADD_BEFORE_INFLATION)
}
```

First, if this was correct, it would make sense to check the last one byte to be `0x00`, as I understood after answering to you and debugging the code. Second, it just does not work unfortunately, the data gets corrupted this way.

yschimke commented:
@AntonRutkevich do you want to work this PR through the kotlin migration?

AntonRutkevich commented:
@yschimke yep, I do.

AntonRutkevich commented:
Sorry for the delay.

Please take a look at the rebased (and migrated to Kotlin) version of this PR. 

There are 2 conceptual issue which I'm not sure how to solve. Mentioned them inlined in the code.

AntonRutkevich commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -56,7 +56,9 @@ class RealWebSocket(
-  private val pingIntervalMillis: Long
+  private val pingIntervalMillis: Long,
+  compression: Boolean = false,
+  contextTakeover: Boolean = false
```
First issue here. 
While we enable compression by default, this does not work well for tests, as they have lots of golden values relying on the compression being off. For now, compression options are in OFF state on `RealWebSocket` creation, but are **enabled** by default in the `RealWebSocket.connect` method (which is only used by "client" mode).

This way, in "client" mode, RealWebSocket has compression enabled by default, but in "server" mode compression is disabled by default. Looks like "server" mode is not exposed in the API and is for tests only, so I think it's acceptable.

@swankjesse please share your thoughts/suggestions.

AntonRutkevich replied:
@swankjesse , is there anything else I can do here?

AntonRutkevich commented on okhttp/src/test/java/okhttp3/internal/ws/RealWebSocketTest.java:
```diff
@@ -350,6 +351,63 @@
+    server.initWebSocket(random, 0, false, false);
+
+    server.webSocket.send(ByteString.encodeUtf8("Hello"));
+    server.webSocket.writeOneFrame$okhttp();
```
Second issue here. 

The 4 tests verify that the actual data is compressed `if (contextTakeover || compressedDataSize < rawDataSize)`. 
Therefore, we need to see the data from the buffer _after_ it is written and _before_ it is read. Calling `writeOneFrame` manually is the only way I found that catches the data in this intermediate state. 

The (minor?) issue with this is that IntelliJ raises a soft error (IDE-level, compilation not affected): 
![Screen Shot 2019-07-19 at 19 39 00](https://user-images.githubusercontent.com/14860034/61551106-efe53580-aa5c-11e9-83bb-a7113ad5e5b6.png)

Not sure if it's a bug or not, but I think that it should not be a blocker.

avently commented:
@AntonRutkevich hello. Did you fix the issue we talked about? I use your old code with try-catch block for now.

AntonRutkevich commented:
Hi @avently, 

Not really. Looks like it goes down to the implementation of the java inflater and requires deeper investigation on how deflate algorithm actually works. 
For now I thinks that try-catch is an acceptable solution, taking into account that inflater does not collapse after this exception. But we definitely need an approve from Jesse for that.

Do you have other ideas?

ming13 commented:
@swankjesse, hey, sorry to ping you about this but can we ask you to take a look at the PR once again? It has been more than four months since it was submitted and it is kind of stuck in the limbo. We would like to make it happen since it will allow us to use WebSockets more efficiently (we actually pay for user data usage in one of the applications we have). Maybe we can make a call with @AntonRutkevich to discuss further steps and how you imagine the implementation should look in the end. We understand that this change is not a trivial one and requires proper thinking about consequences.

avently commented:
>Do you have other ideas?

Nope. Works ok for me with try-catch. Maybe someone who knows inflater better has answers. But this guy probably saw this discussion here and in okio thread. No suggestions were made so let's leave it with try-catch. It doesn't ruin anything

swankjesse commented:
@ming13 yeah, this is ready for attention. Now that the dust around our Kotlin migration has recently settled, I’m ready.

AntonRutkevich commented:
Hey @swankjesse , sorry for the delay. 

Could you please confirm what needs to be done from my side to get this PR merged?

yschimke commented:
@swankjesse ping - any thoughts?

AntonRutkevich commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -303,13 +327,19 @@ class RealWebSocket(
+        readerToClose = this.reader
+        this.reader = null
+        deflaterToClose = this.messageDeflater
+        this.writer = null
```
Indeed, thanks! Fixed

AntonRutkevich commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -433,6 +467,10 @@ class RealWebSocket(
+            readerToClose = this.reader
+            this.reader = null
+            deflaterToClose = this.messageDeflater
+            this.writer = null
```
Fixed

AntonRutkevich commented:
@swankjesse , fixed the comments, the 2 failing `ci/circleci: testjdk11` tests are 
```
okhttp3.CallTest > recoverFromTlsHandshakeFailure FAILED
    org.junit.runners.model.TestTimedOutException: test timed out after 30000 milliseconds
okhttp3.CallTest > tooManyProxyAuthFailuresWithConnectionClose FAILED
    org.junit.runners.model.TestTimedOutException: test timed out after 30000 milliseconds
```
which, I believe, are flaky and are outside of the PR scope.

yschimke commented:
Builds are green, think it's just the merge conflict blocking this now.

avently commented:
@AntonRutkevich yesterday I saw a video made by you about testing a code. Thank you.
I thought, I saw Anton Rutkevich somewhere. It turns out I use your build since June.

AntonRutkevich commented:
@yschimke  Done. Had to move the diff of #5800 by @dave-r12 from `RealWebSocketTest.java` to `RealWebSocketTest.kt`.
Not sure how to preserve Dave's authorship, as the `.java` file is gone now.

@avently, glad it was useful :)

avently commented:
@AntonRutkevich 
Does the latest code have a fix for situation we talked earlier? 
https://github.com/square/okhttp/pull/4871#issuecomment-498684696

For those who curious how much effective this pr is:
it reduces bandwidth of my app by 85.5%. It's alot!

AntonRutkevich commented:
@avently 
No, the current code does not have the fix for the above situation. The reason is that the only way to fix it completely I could find was the ugly try-catch from https://github.com/square/okhttp/pull/4871#issuecomment-499140329.

It's great to see such traffic reduction!

avently commented:
@AntonRutkevich 
So
1. Ugly but working for a half of year without any problems
2. Good but crashes every time on two leading crypto exchanges from two.

And you chose 1 option? I think we should have a try-catch block because it proved it's usefulness in real world application

AntonRutkevich commented:
I'm fine with 1) as long as I get an approve from @swankjesse for that :)

avently commented:
I built a code from this PR and the problem is still here... Just to confirm.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -100,6 +108,13 @@ class RealWebSocket(
+   * Actual will be defined based on the server response.
+   * For server, actual compression options for outbound stream.
+   */
+  private var options: WebSocketOptions = WebSocketOptions(compression, contextTakeover)
```
I think the term of art for Web Sockets is extensions, not options

swankjesse replied:
These parameters are a slight simplification of the permessage-deflate spec. In particular...

contextTakeover is independent for client-sent and server-sent messages
there’s `client_max_window_bits` and `server_max_window_bits` options

We need to detect and recognize these options and reject negotiations that we cannot or do not want to implement. Most particularly, we cannot implement `client_max_window_bits` for any value other than 15.

swankjesse replied:
(we can, however, operate just fine if that option is absent).

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -147,6 +162,11 @@ class RealWebSocket(
         .header("Sec-WebSocket-Version", "13")
+        .apply {
+          if (originalRequest.header(HEADER_WS_EXTENSION) == null) {
+            header(HEADER_WS_EXTENSION, HEADER_VALUE_CONTEXT_TAKEOVER)
```
I think we need an else clause here. If the extension is already present, we need to do one of two things:

 - disable our code to handle compression
 - fail outright

My preference is to fail outright; perhaps very early. Unlike HTTP compression, there’s no way the application layer can effectively implement WebSocket extensions on top of OkHttp’s APIs. This is because they don’t get access to the reserved bits.

So - - - I propose we change OkHttp to explode if the WebSocket Request has a `HEADER_WS_EXTENSION` set.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -81,6 +83,12 @@ class RealWebSocket(
+   * Null until web socket is connected with compression enabled.
+   * Used for compressing data at message level.
+   */
+  private var messageDeflater: MessageDeflater? = null
```
I think it’s weird that this is on the RealWebSocket but the inflater is on the MessageReader. Seems needlessly asymmetric.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -156,6 +176,7 @@ class RealWebSocket(
         try {
           checkUpgradeSuccess(response, exchange)
           streams = exchange!!.newWebSocketStreams()
+          options = WebSocketOptions.parseServerResponse(response)
```
I don’t love that this field gets reassigned. Creating two fields for offered and agreed would be a nice improvement I think.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -305,13 +329,19 @@ class RealWebSocket(
 
     var toClose: Streams? = null
+    var readerToClose: WebSocketReader? = null
+    var deflaterToClose: MessageDeflater? = null
```
yep, I think this is too asymmetric

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -324,6 +354,8 @@ class RealWebSocket(
     } finally {
       toClose?.closeQuietly()
+      readerToClose?.closeQuietly()
+      deflaterToClose?.closeQuietly()
```
Really nice attention to closing stuff. Thank you!

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -457,13 +495,27 @@ class RealWebSocket(
+          val deflated = deflater.deflate(uncompressedData).readByteString()
+          // If context takeover is NOT used, it's guaranteed that sending uncompressed message
+          // which is smaller than compressed version will be more efficient.
+          // We cannot make such assumptions if context takeover is used.
```
Good comment.

swankjesse replied:
For simplicity I’d like to pay the ~14-18 bytes overhead and remove this check. I expect that everyone is going to use context takeover always, and so it’s complexity in our code to save 18 bytes that will probably never exist.

swankjesse replied:
(According to RFC 1979, the deflater won’t add more than 14-18 bytes of overhead on data that doesn’t compress well)

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt:
```diff
@@ -566,5 +628,15 @@ class RealWebSocket(
+    /**
+     * Java Deflater does not allow specifying window size
+     * and the underlying zlib uses default value of MAX_WBITS=15.
+     * It's the default window size for many deflate implementations too.
```
Oh sweet, we’re reading the same source code. I think we might get better interoperability by omitting the `server_max_window_bits` and `client_max_window_bits` parameters, since 15 bits is the default?

> Absence of this extension parameter in an extension negotiation
> response indicates that the server can receive messages compressed
> using an LZ77 sliding window of up to 32,768 bytes.

swankjesse replied:
Yeah, I think for correctness we actually need to remove these parameters:
https://tools.ietf.org/html/rfc7692#section-7.1.2.2

I think if we include a value, the server can agree on a lower value. But we don’t have a mechanism to use a smaller window bits size through the Java APIs. So our server would have say, a 4 KiB inflate window and the client would have a 32 KiB deflate window and we’d be in trouble.

Omitting the parameter altogether prevents the problem because the server is forbidden from returning a value for the parameter if the client doesn’t mention it.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+      }
+
+      val extension = header
+        .split(", ")
```
whitespace is optional after commas

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+        throw ProtocolException("$HEADER_WEB_SOCKET_EXTENSION malformed: '$header'")
+      }
+
+      val tokens = extension.split("; ").map { it.replace(";", "") }
```
more whitespace that must be optional

swankjesse replied:
also, we’ve got code in Util.kt that can decode this with fewer allocations

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+        // See https://tools.ietf.org/html/rfc7692#section-5.2 for details.
+        .getOrNull(0)
+
+      if (extension.isNullOrBlank()) {
```
why would it be null?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+        .split(", ")
+        // Additional or fallback extensions are not supported currently.
+        // See https://tools.ietf.org/html/rfc7692#section-5.2 for details.
+        .getOrNull(0)
```
Why would it be null? Splitting a string always returns >= 1 result.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+      }
+
+      val tokens = extension.split("; ").map { it.replace(";", "") }
+      if (tokens.isEmpty()) {
```
Splitting a string always returns >= 1 result

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+        throw ProtocolException("Duplicate options found in '$extension'")
+      }
+
+      val clientNoContextTakeover = Option.CLIENT_NO_CONTEXT_TAKEOVER in options
```
There’s also an independent field, server_no_context_takeover, that we need to handle when inflating

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+    /**
+     * If server decides to use MAX_WINDOW_BITS less than 15, is fine.
+     * The deflater has enough information from the deflated data
+     * to parse it correctly.
```
Unfortunately, if the server says client_max_window_bits is fewer than 14, the server won‘t be able to read what we write.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+
+    private const val SUPPORTED_CLIENT_MAX_WINDOW_BITS = MAX_MAX_WINDOW_BITS
+
+    fun parse(option: String): Option =
```
I think we’ll be happier if we never offer the `client_max_window_bits` or `server_max_window_bits` parameters, which makes them forbidden in the agreement response.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt:
```diff
@@ -0,0 +1,158 @@
+        else -> throw ProtocolException("Unknown option: '$option'")
+      }
+
+    private fun verifyMaxWindowBits(option: Option, optionString: String) {
```
... and then we can delete this!

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/RealWebSocketTest.kt:
```diff
@@ -0,0 +1,495 @@
+import org.junit.Ignore
+import org.junit.Test
+
+class RealWebSocketTest {
```
Converting this to Kotlin is good! Though we should split that into a separate PR to land.

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketOptionsTest.kt:
```diff
@@ -0,0 +1,237 @@
+
+  private companion object {
+
+    val NO_COMPRESSION = WebSocketOptions(
```
I’ve seen these before

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketOptionsTest.kt:
```diff
@@ -0,0 +1,237 @@
+     */
+    @JvmStatic
+    @Parameterized.Parameters(name = "{0}")
+    fun parameters(): Collection<Array<out Any?>> = listOf(
```
You might be able to make each of these parameters into its own test? I don’t think it needs to be NxM?

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketReaderTest.java:
```diff
@@ -51,29 +57,54 @@
-    data.clear();
+  }
+
+  @Test public void reservedFlag1IsUnsupportedForControlFrames() throws IOException {
```
Nice

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketReaderTest.java:
```diff
@@ -51,29 +57,54 @@
+    }
+  }
+
+  @Test public void reservedFlag1IsUnsupportedForContinuationFrames() throws IOException {
```
Nice

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketReaderTest.java:
```diff
@@ -113,12 +144,36 @@
+  }
+
+  @Test public void clientWithCompressionSimpleCompressedHello() throws IOException {
+    data.write(ByteString.decodeHex("c107f248cdc9c90700")); // Hello
```
oh sweet, that’s this one: https://tools.ietf.org/html/rfc7692#section-7.2.3.1
       0xf2 0x48 0xcd 0xc9 0xc9 0x07 0x00 0x00 0x00 0xff 0xff

swankjesse replied:
We should link that

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketReaderTest.java:
```diff
@@ -182,6 +283,15 @@
+    try {
+      clientReaderWithCompression.processNextFrame();
+      fail();
+    } catch (IOException ignored) {
```
ignored -> expected ?

swankjesse commented on okhttp/src/test/java/okhttp3/internal/ws/WebSocketReaderTest.java:
```diff
@@ -315,6 +425,17 @@
+    try {
+      clientReaderWithCompression.processNextFrame();
+      fail();
+    } catch (IllegalStateException e) {
```
e -> expected

AntonRutkevich commented:
Hi @avently, 

you might want to check if the Jesse's implementation in #5871 still has the EOF issue on you input. It differs a lot in the details from this PR (much simpler also), so the issue might be gone there.

avently commented:
@AntonRutkevich thank you for letting me know. I will read the changes you mentioned. Glad that your work received continuation.
